src/plugins/profiling/common/functions.ts

```diff
@@ -0,0 +1,126 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import {
+  compareFrameGroup,
+  defaultGroupBy,
+  Executable,
+  FileID,
+  FrameGroup,
+  FrameGroupID,
+  groupStackFrameMetadataByStackTrace,
+  hashFrameGroup,
+  StackFrame,
+  StackFrameID,
+  StackFrameMetadata,
+  StackTrace,
+  StackTraceID,
+} from './profiling';
+
+type TopNFunctionAndFrameGroup = {
+  Frame: StackFrameMetadata;
+  FrameGroup: FrameGroup;
+  CountExclusive: number;
+  CountInclusive: number;
+};
+
+type TopNFunction = Pick<TopNFunctionAndFrameGroup, 'Frame' | 'CountExclusive' | 'CountInclusive'>;
+
+type TopNFunctions = {
+  TotalCount: number;
+  TopN: TopNFunction[];
+};
+
+export function createTopNFunctions(
+  events: Map<StackTraceID, number>,
+  stackTraces: Map<StackTraceID, StackTrace>,
+  stackFrames: Map<StackFrameID, StackFrame>,
+  executables: Map<FileID, Executable>,
+  startIndex: number,
+  endIndex: number
+): TopNFunctions {
+  const metadata = groupStackFrameMetadataByStackTrace(stackTraces, stackFrames, executables);
+
+  // The `count` associated with a frame provides the total number of
+  // traces in which that node has appeared at least once. However, a
+  // frame may appear multiple times in a trace, and thus to avoid
+  // counting it multiple times we need to record the frames seen so
+  // far in each trace.
+  let totalCount = 0;
+  const topNFunctions = new Map<FrameGroupID, TopNFunctionAndFrameGroup>();
+
+  // Collect metadata and inclusive + exclusive counts for each distinct frame.
+  for (const [traceHash, count] of events) {
+    const uniqueFrameGroupsPerEvent = new Set<FrameGroupID>();
+
+    totalCount += count;
+
+    const frames = metadata.get(traceHash)!;
+    for (let i = 0; i < frames.length; i++) {
+      const frameGroup = defaultGroupBy(frames[i]);
+      const frameGroupID = hashFrameGroup(frameGroup);
+
+      if (!topNFunctions.has(frameGroupID)) {
+        topNFunctions.set(frameGroupID, {
+          Frame: frames[i],
+          FrameGroup: frameGroup,
+          CountExclusive: 0,
+          CountInclusive: 0,
+        } as TopNFunctionAndFrameGroup);
+      }
+
+      const topNFunction = topNFunctions.get(frameGroupID)!;
+
+      if (!uniqueFrameGroupsPerEvent.has(frameGroupID)) {
+        uniqueFrameGroupsPerEvent.add(frameGroupID);
+        topNFunction.CountInclusive += count;
+      }
+
+      if (i === 0) {
```

Comment:
We reversed the order of frames in the backend. Do we reverse it again when parsing the query results ? (If not, this should be `if (i === frames.length - 1)`.)

Reply:
Thanks, the stack frames used to build the TopN functions response was using the leaf-frame-last ordering so I fixed it as you suggested.